classes/task/merge_user_accounts.php

```diff
@@ -0,0 +1,198 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+namespace tool_mergeusers\task;
+defined('MOODLE_INTERNAL') || die();
+require_once(__DIR__ . '/../../lib/autoload.php');
+use MergeUserTool;
+use \tool_mergeusers\merge_request;
+/**
+ * Adhoc task to merge user accounts.
+ *
+ * @package   tool_mergeusers
+ * @author    Jordi Pujol-Ahulló <jordi.pujol@urv.cat>
+ * @author    Nicola Vallinoto <n.vallinoto@liguriadigitale.it>
+ * @copyright 2013 Servei de Recursos Educatius (http://www.sre.urv.cat)
+ * @copyright 2023 Liguria Digitale (http://www.liguriadigitale.it)
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * */
+class merge_user_accounts extends \core\task\adhoc_task {
+    /**
+     * Execute the task of merge users accounts.
+     */
+    public function execute() {
+        global $DB;
+        $data = $this->get_custom_data();
+        $maxattempts = get_config('tool_mergeusers', 'maxattempts');
+        $mergerequest = $DB->get_record(merge_request::TABLE_MERGE_REQUEST,
+                                            ['id' => $data->mergerequestid]);
+        if (!$mergerequest) {
+            // Merge request was removed before running this task.
+            return;
+        }
+        $this->verify_users_to_keep_and_remove($mergerequest);
+        $mergerequestresult = $this->merge($mergerequest, $maxattempts, merge_request::TRIED_WITH_ERROR);
+        if ($mergerequestresult->status == merge_request::COMPLETED_WITH_SUCCESS) {
+            /* We run the merge request AGAIN because the user may be interacting with Moodle
+            * while merge request is being processed, so that NO ALL records are correctly migrated
+            * into the user to keep. It will ensure ALL records are correctly migrated into the user to keep. */
+            $mergerequestresult = $this->merge($mergerequestresult, $maxattempts, merge_request::COMPLETED_WITH_ERRORS);
+        }
+        if ($mergerequestresult->status != merge_request::COMPLETED_WITH_ERRORS &&
+            $mergerequestresult->status != merge_request::COMPLETED_WITH_SUCCESS) {
+            // Throwing exception will ensure this adhoc task is re-queued until $maxretries is reached.
+            throw new moodle_exception(get_string('failedmergerequest', 'tool_mergeusers'));
```

Comment:
I would also add the current `$mergerequest->id` and `$mergerequest->retries` into the message of the exception, into the `failedmergerequest` string. This way, we provide enough context into the exception message and int he trace log.

Reply:
done

Reply:
I don't the the parameters ont the get_string() yet.

Reply:
done

Reply:
revisit the API of the constructor of [moodle_exception](https://github.com/moodle/moodle/blob/e77452220b268f70612b9b915179c17fc6d93173/lib/setuplib.php#L100), `function __construct($errorcode, $module='', $link='', $a=NULL, $debuginfo=null) {`

the instantiation should be something like:

`throw new moodle_exception('failedmergerequest', 'tool_mergeusers');`

and if this, imagine, has to have some parameters on the string, it should be called like this:

`throw new moodle_exception('failedmergerequest', 'tool_mergeusers', '', $a);`

Revisit all moodle_exception to be invoked this way.